mne_connectivity/decoding/decomposition.py

```diff
@@ -0,0 +1,474 @@
+            verbose, (bool, str, int, None), "`verbose`", "bool, str, int, or None"
+        )
+
+        # Store inputs
+        self.info = info
+        self.fmin = fmin
+        self.fmax = fmax
+        self._indices = _indices  # uses getter/setter for public parameter
+        self.mode = mode
+        self.mt_bandwidth = mt_bandwidth
+        self.mt_adaptive = mt_adaptive
+        self.mt_low_bias = mt_low_bias
+        self.cwt_freq_resolution = cwt_freq_resolution
+        self.cwt_n_cycles = cwt_n_cycles
+        self.n_components = 1  # XXX: fixed until n_comps > 1 supported
+        self._rank = _rank  # uses getter/setter for public parameter
+        self.n_jobs = n_jobs
+        self.verbose = verbose
+
+    def _check_indices(self, indices, n_chans):
+        """Check that the indices input is valid."""
+        # convert to multivariate format and check validity
+        indices = _check_multivariate_indices(([indices[0]], [indices[1]]), n_chans)
+
+        # find whether entries of indices exceed number of channels
+        max_idx = np.max(indices.compressed())
+        if max_idx + 1 > n_chans:
+            raise ValueError(
+                "At least one entry in `indices` is greater than the number of "
+                "channels in `info`"
+            )
+
+        return indices
+
+    def _check_rank(self, rank, indices):
+        """Check that the rank input is valid."""
+        if rank is not None:
+            # convert to multivariate format
+            rank = ([rank[0]], [rank[1]])
+
+            # find whether entries of rank exceed number of channels in indices
+            if rank[0][0] > len(indices[0]) or rank[1][0] > len(indices[1]):
+                raise ValueError(
+                    "At least one entry in `rank` is greater than the number of "
+                    "seed/target channels in `indices`"
+                )
+
+        return rank
+
+    def fit(self, X, y=None):
+        """Compute connectivity decomposition filters for epoched data.
+
+        Parameters
+        ----------
+        X : array, shape=(n_epochs, n_signals, n_times)
+            The input data which the connectivity decomposition filters should be fit
+            to.
+        y : None
+            Used for scikit-learn compatibility.
+
+        Returns
+        -------
+        self : instance of CaCoh | MIC
+            The modified class instance.
+        """
+        # validate input data
+        self._check_X(X, ndim=[3])
+        self._get_rank_and_ncomps_from_X(X)
+
+        # compute CSD
+        csd = self._compute_csd(X)
+
+        # instantiate connectivity estimator and add CSD information
+        self._conn_estimator = self._conn_estimator(
+            n_signals=X.shape[1],
+            n_cons=1,
+            n_freqs=1,
+            n_times=0,
+            n_jobs=self.n_jobs,
+            store_filters=True,
+        )
+        self._conn_estimator.accumulate(con_idx=np.arange(csd.shape[0]), csd_xy=csd)
```

Comment:
here again we assume that the connectivity estimater operates on the CSD. Is there a way to refactor this so that `_AbstractDecompositionBase.fit` is a more general and `_AbstractCoherenceDecomposition.fit` actually uses the CSD?

Reply:
As I mentioned above I think it would be better to have something like `_AbstractSpectralDecomposition` than `_AbstractCoherenceDecomposition` where CSD-based methods could be handled.

In this case the base `fit` could cover checking the input data and its rank. The question is whether the connectivity estimator should also be instantiated in the base class. Would we expect all decomposition methods to have an equivalent connectivity estimator (i.e. something that could also be used with `spectral_connectivity_epochs/time`)?

Then like you say, computing CSD and adding it to the connectivity estimator (or whatever alternative there is) could be handled in the daughter class. Maybe a `_fit` method is called?

Coming back to the base class, the filters and patterns could then be computed and extracted. This should be okay to do in the base class as long as the connectivity estimator/alternative use the same method/attribute names, e.g. `compute_con()` for computing the filters and patterns which are stored as `.filters` and `.patterns` attributes.